contracts/RewardDistributor.sol

```diff
@@ -81,404 +97,583 @@ contract RewardDistributor is RewardDistributorStorage, Exponential {
-
-        if (borrowerIndex.mantissa > 0) {
-            Double memory deltaIndex = sub_(borrowIndex, borrowerIndex);
-            uint256 borrowerAmount = div_(JToken(jToken).borrowBalanceStored(borrower), marketBorrowIndex);
-            uint256 borrowerDelta = mul_(borrowerAmount, deltaIndex);
-            uint256 borrowerAccrued = add_(rewardAccrued[rewardType][borrower], borrowerDelta);
-            rewardAccrued[rewardType][borrower] = borrowerAccrued;
-            emit DistributedBorrowerReward(rewardType, JToken(jToken), borrower, borrowerDelta, borrowIndex.mantissa);
+        uint256 marketBorrowIndex
+    ) private verifyRewardType(rewardType) returns (uint256) {
+        uint256 borrowIndex = rewardBorrowState[rewardType][jToken].index;
+        uint256 borrowerIndex = rewardBorrowerIndex[rewardType][jToken][borrower];
+
+        uint256 deltaIndex = borrowIndex.sub(borrowerIndex);
+        uint256 borrowerAmount = JToken(jToken).borrowBalanceStored(borrower).mul(expScale).div(marketBorrowIndex);
+        uint256 borrowerReward = borrowerAmount.mul(deltaIndex).div(doubleScale);
+
+        if (borrowIndex != borrowerIndex) {
+            rewardBorrowerIndex[rewardType][jToken][borrower] = borrowIndex;
         }
+        emit DistributedBorrowerReward(rewardType, JToken(jToken), borrower, borrowerReward, borrowIndex);
+        return borrowerReward;
     }
 
     /**
-     * @notice Refactored function to calc and rewards accounts supplier rewards
-     * @param jToken The market to verify the mint against
-     * @param supplier The supplier to be rewarded
+     * @notice Claim all JOE/AVAX accrued by the holders
+     * @param rewardType 0 = JOE, 1 = AVAX
+     * @param holder The address to claim JOE/AVAX for
+     * @param jTokens The list of markets to claim JOE/AVAX in
+     * @param borrower Whether or not to claim JOE/AVAX earned by borrowing
+     * @param supplier Whether or not to claim JOE/AVAX earned by supplying
      */
-    function updateAndDistributeSupplierRewardsForToken(address jToken, address supplier) external {
-        require(adminOrInitializing(), "only admin can update and distribute supplier rewards");
-        for (uint8 rewardType = 0; rewardType <= 1; rewardType++) {
-            updateRewardSupplyIndex(rewardType, jToken);
-            distributeSupplierReward(rewardType, jToken, supplier);
+    function _claimReward(
+        uint8 rewardType,
+        address payable holder,
+        JToken[] memory jTokens,
+        bool borrower,
+        bool supplier
+    ) private verifyRewardType(rewardType) {
+        uint256 rewards = rewardAccrued[rewardType][holder];
+        if (!claimedFromOldRewarder[rewardType][holder]) {
+            // claim from previous rewarder
+            oldRewarder.claimReward(rewardType, holder);
+            rewards = rewards.add(oldRewarder.rewardAccrued(rewardType, holder));
+            claimedFromOldRewarder[rewardType][holder] = true;
+        }
```

Comment:
Yup the idea is to empty the rewards and set the speed to 0, so all accrued rewards are sent to their respective `rewardAccrued` and added to the new rewarder's one.
This will allow to update the rewarder without any user having to do anything to get their previous reward or future ones.